hw/model/src/model_emulated.rs

```diff
@@ -0,0 +1,342 @@
+use crate::McuHwModel;
+use crate::Output;
+use anyhow::Result;
+use caliptra_emu_bus::Clock;
+use caliptra_emu_bus::Event;
+use caliptra_emu_cpu::{Cpu as CaliptraCpu, InstrTracer as CaliptraInstrTracer};
+use caliptra_emu_periph::ActionCb;
+use caliptra_emu_periph::MailboxRequester;
+use caliptra_emu_periph::ReadyForFwCb;
+use caliptra_emu_periph::{CaliptraRootBus, CaliptraRootBusArgs, TbServicesCb};
+use caliptra_hw_model::ModelError;
+use caliptra_hw_model_types::ErrorInjectionMode;
+use caliptra_image_types::IMAGE_MANIFEST_BYTE_SIZE;
+use emulator_bus::BusConverter;
+use emulator_bus::Clock as McuClock;
+use emulator_cpu::{Cpu as McuCpu, InstrTracer as McuInstrTracer, Pic};
+use emulator_periph::{I3c, I3cController, Mci, McuRootBus, McuRootBusArgs, Otp};
+use emulator_registers_generated::root_bus::AutoRootBus;
+use std::cell::Cell;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::Hasher;
+use std::io::Write;
+use std::path::PathBuf;
+use std::rc::Rc;
+use std::sync::mpsc;
+
+/// Emulated model
+pub struct ModelEmulated {
+    caliptra_cpu: CaliptraCpu<BusLogger<CaliptraRootBus>>,
+    mcu_cpu: McuCpu<McuBusLogger<AutoRootBus>>,
+    output: Output,
+    caliptra_trace_fn: Option<Box<CaliptraInstrTracer<'static>>>,
+    mcu_trace_fn: Option<Box<McuInstrTracer<'static>>>,
+    ready_for_fw: Rc<Cell<bool>>,
+    cpu_enabled: Rc<Cell<bool>>,
+    trace_path: Option<PathBuf>,
+
+    // Keep this even when not including the coverage feature to keep the
+    // interface consistent
+    _rom_image_tag: u64,
+    iccm_image_tag: Option<u64>,
+
+    events_to_caliptra: mpsc::Sender<Event>,
+    events_from_caliptra: mpsc::Receiver<Event>,
+    collected_events_from_caliptra: Vec<Event>,
+}
+
+fn hash_slice(slice: &[u8]) -> u64 {
+    let mut hasher = DefaultHasher::new();
+    std::hash::Hash::hash_slice(slice, &mut hasher);
+    hasher.finish()
+}
+
+impl McuHwModel for ModelEmulated {
+    fn new_unbooted(params: InitParams) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        let clock = Clock::new();
+        let timer = clock.timer();
+
+        let ready_for_fw = Rc::new(Cell::new(false));
+        let ready_for_fw_clone = ready_for_fw.clone();
+
+        let cpu_enabled = Rc::new(Cell::new(false));
+        let cpu_enabled_cloned = cpu_enabled.clone();
+
+        let output = Output::new(params.log_writer);
+
+        let output_sink = output.sink().clone();
+
+        let bus_args = CaliptraRootBusArgs {
+            rom: params.caliptra_rom.into(),
+            tb_services_cb: TbServicesCb::new(move |ch| {
+                output_sink.set_now(timer.now());
+                output_sink.push_uart_char(ch);
+            }),
+            ready_for_fw_cb: ReadyForFwCb::new(move |_| {
+                ready_for_fw_clone.set(true);
+            }),
+            bootfsm_go_cb: ActionCb::new(move || {
+                cpu_enabled_cloned.set(true);
+            }),
+            security_state: params.security_state,
```

Comment:
A lot of this isn't necessary. I removed a lot of it, and can remove more as we build out the models.